### DISCUSSION TITLE
Convert geometry from root to Acts

### DISCUSSION THREAD

**Discussion by pbelecky** [acts-project/acts#1330] [General]
Hello.
I'm new to Acts, and I have a question.
I need to convert the geometry of the detector, which is in the root file, to Acts.

To do this, I'm trying to use an example from [
Tests/UnitTests/Plugins/TGeoParserTests.cpp](https://github.com/acts-project/acts/blob/main/Tests/UnitTests/Plugins/TGeo/TGeoParserTests.cpp)

But it explicitly specifies the targetname list on [line 42](https://github.com/acts-project/acts/blob/main/Tests/UnitTests/Plugins/TGeo/TGeoParserTests.cpp#L42), as I understand it, this is a list of leaf TGeoVolume from the root geometry:
```
tgpOptions.targetNames = {"PixelActiveo2", "PixelActiveo4", "PixelActiveo5",
"PixelActiveo6"};
```

But in my geometry there are several hundred leaf TGeoVolume s , and specify them manually in TGeoParser::Options::targetNames is very inconvenient.

How do I convert the entire geometry from root to Acts?

What is the common way to do this?

What should I read about it.

Thanks for the answers!

**Comment by paulgessinger** [comment]
Hey @pavel532!

In principle, the point of these name based selectors is that you don't want to convert the **entire** geometry to ACTS, but only the sensitive elements. Passive structures are treated differently, and are generally not modelled explicitly in the geometry. Unfortunately, there's no good way to determine automatically what the sensitive elements are in the TGeo hierarchy.

If I remember correctly, `tgpOptions.targetNames` accepts pattern strings with `*` to for example select `PixelActiveo*` in your example here, which might make this more manageable.

Let me also add some people who have some experience with the TGeo conversion: @niermann999 @osbornjd.

**Comment by asalzburger** [comment]
If you want (and can) share your geometry file + a bit of description, we might be able to help you with the setup.

**Comment by pbelecky** [comment]
Thanks @paulgessinger!
I tried wildcards - `"*"` doesn't work, but like `"a*"` it works.
This helped me.
Test example from Acts panda.root - I managed to convert the whole thing.
But now the following problems have arisen.

@asalzburger ,
I use [MpdRoot](https://git.jinr.ru/nica/mpdroot).
It is based on [FairRoot](https://github.com/FairRootGroup/FairRoot).

Here is the root file with the geometry that I am trying to convert:
https://git.jinr.ru/nica/mpdroot/-/blob/dev/geometry/tpc_v9.root

To do this, I installed MpdRoot, linked it with Acts.
And to convert, I use the following code:
```c++
GeometryContext tgContext = GeometryContext();
TGeoParser::Options tgpOptions;
std::string volumeName = "*"; 
tgpOptions.volumeNames = {volumeName};
tgpOptions.targetNames = {"TPC*", "tpc*"};
std::string axes = "XY*";
TGeoParser::State tgpState;
TFile f("tpc_v9.root", "READ");
TGeoVolumeAssembly *geoVolA;
f.GetObject("TOP", geoVolA);
TGeoVolume *vol = geoVolA->GetNode(0)->GetMotherVolume();
tgpState.volume = vol;

TGeoParser::select(tgpState, tgpOptions);

ObjVisualization3D objVis;
for (auto& snode : tgpState.selectedNodes) {
    const auto& shape = *(snode.node->GetVolume()->GetShape());
    const auto& transform = *(snode.transform.get());
    auto surface =
        TGeoSurfaceConverter::toSurface(shape, transform, axes, scale);
    GeometryView3D::drawSurface(objVis, *surface, tgContext);
}

objVis.write("PixelActive");
```

**Comment by paulgessinger** [comment]
> I tried wildcards - `"*"` doesn't work, but like `"a*"` it works.

The pattern implementation is fairly simplistic. Maybe this wasn't implemented. We could expand the mechanism if needed.

**Comment by pbelecky** [comment]
I converted the Root geometry of the TPC into a sequence of ACTS::Surface-s using TGeoSurfaceConverter::toSurface().

How do I get TrackingGeometry from them?

**Reply by paulgessinger** [reply]
It's possible you can make one of the automated geometry builder (e.g. cylinder) combine these surfaces into a tracking geometry automatically.

If that doesn't work, you'll likely have to create layer instances from these surfaces yourself (e.g. via one of the layer creators) with a custom layer builder that you can configure in a tracking volume builder, which in turn can be plugged into a tracking geometry builder.

Hope that helps!

**Reply by osbornjd** [reply]
For reference, in sPHENIX we chose a different approach. We made planar surfaces that approximated a cylinder, each planar surface spanning 3 degrees in azimuth. These are our "fake surfaces" which are picked up by the Acts geometry building code and automatically added to the tracking geometry. The planar surfaces mock up a cylindrical surface which approximates the 48 layers of our TPC. In this approach we explicitly modify the TGeoManager object to make the TPC active surfaces these fake planar surfaces.
